Revise the section on required / optional / nullable fields. (#6468)
pydantic/pydantic: docs/migration.md

#### Required, optional, and nullable fields

Pydantic V2 changes some of the logic for specifying whether a field annotated as `Optional` is required

> (i.e., has no default value) or not (i.e., has a default value of `None`), and now more closely matches the

Reviewer: ```suggestion
(i.e., has no default value) or not (i.e., has a default value of `None` or any other value of the corresponding type), and now more closely matches the
```

behavior of `dataclasses`. Similarly, fields annotated as `Any` no longer have a default value of `None`.

The following table describes the behavior of field annotations in V2:

| State                                            | Field Definition           |
|--------------------------------------------------|----------------------------|
| Required, cannot be `None`                       | `f1: str`                  |

> | Required, can be `None`                          | `f2: Optional[str]`        |
> | Not required, can be `None`                      | `f3: Optional[str] = None` |

Reviewer: I feel like this could give a slightly incorrect impression that exactly `= None` is needed to transform a required field into a not required one.

I think that it is worth mentioning here that any default value if provided makes a field not required.

Something like this maybe
```suggestion
| Not required, cannot be `None`, is `'abc'` by default | `f3: str = 'abc'`      |
| Required, can be `None`                               | `f2: Optional[str]`      |
| Not required, can be `None`, is `None` by default     | `f3: Optional[str] = None`  |
| Not required, can be `None`, is `'abc'` by default    | `f3: Optional[str] = 'abc'` |
```

#### Required, optional, and nullable fields

> Pydantic V2 introduces stricter definitions for specifying if a field is `required` or `optional` – more closely
> matching the behavior `dataclasses`.

Reviewer 2: ```suggestion
Pydantic V2 changes some of the logic for specifying whether a field annotated as `Optional` is required
(i.e., has no default value) or not (i.e., has a default value of `None`), and now more closely matches the
behavior of `dataclasses`.
```

> In V2, fields are specified as follows:

Reviewer 2: ```suggestion
The following table describes the behavior of field annotations in V2:
```

| State                        | Field Definition           |
|------------------------------|----------------------------|
| Required, cannot be None     | `f1: str`              |
| Required, can be None        | `f2: Optional[str]`        |
| Not required, can be None    | `f3: Optional[str] = None` |
| Not required, cannot be None | `f4: str = 'Foobar'`       |

| Required | Can be `None` | Field Definition |
|----------|---------------|------------------|
| ✔        | X             | `f1: str`        |
| ✔         | ✔             | `f2: Optional[str]`        |
| X        | ✔             | `f3: Optional[str] = None` |

> | X        | X             | `f4: str = 'Foobar'`       |

Author: Here are 2 formats for expressing field definitions. An additional open would be to show as a bullet point list. (I'd rather have as text in addition to the code example below – since these are breaking changes.)

If we wanted to go overboard, we could show how this is different from V1: 
```
| State                        | V1 Field Definition           | V2 Field Definition           |
|------------------------------|----------------------------|---------------------|
| Required, cannot be None     | `f1: str`              | `f1: str`              |
| Required, can be None        | `f2: Optional[str] = Field(...)`        | `f2: Optional[str]`             |
| Not required, can be None    | `f3: Optional[str]` | `f3: Optional[str]` |
| Not required, cannot be None | `f4: str = 'Foobar'`       | `f4: str = 'Foobar'`       |

```

Reviewer 2: I think the table you have is good as is, don't think we need to expand it to have the "V1 Field Definition"

!!! note
    A field annotated as `typing.Optional[T]` will be required, and will allow for a value of `None`.
    It does not mean that the field has a default value of `None`. _(This is a breaking change from V1.)_

> Code example demonstrating the above:

Reviewer 2: ```suggestion
Here is a code example demonstrating the above:
```
